# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/TrialInlining.cpp
# Commit: dab916754456
# Full Hash: dab916754456373246c85d658fde0520c7246865
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2020-07-23 03:20:25
# Regressor Bug: 1646378
# File Overlap Count: 1
# Description:
#   Bug 1646378: Invalidate the root script in MaybeInvalidateWarp r=jandem
#   
#   When we replace a stub in an inlined ICScript, we want to invalidate the script into which it was inlined, rather than the original copy of the inlined script.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D84431
# ==============================================================================

diff -r 86db033f2b35 -r dab916754456 js/src/jit/TrialInlining.cpp
--- a/js/src/jit/TrialInlining.cpp	Wed Jul 22 11:31:25 2020 +0000
+++ b/js/src/jit/TrialInlining.cpp	Wed Jul 22 21:57:45 2020 +0000
@@ -25,9 +25,13 @@
   ICScript* icScript = frame->icScript();
   bool isRecursive = !!icScript->inliningRoot();
 
-  InliningRoot* root = isRecursive
-                           ? icScript->inliningRoot()
-                           : script->jitScript()->getOrCreateInliningRoot(cx);
+  InliningRoot* root =
+      isRecursive ? icScript->inliningRoot()
+                  : script->jitScript()->getOrCreateInliningRoot(cx, script);
+  if (!root) {
+    return false;
+  }
+
   JitSpew(JitSpew_WarpTrialInlining,
           "Trial inlining for %s script %s:%u:%u (%p) (inliningRoot=%p)",
           (isRecursive ? "inner" : "outer"), script->filename(),
@@ -50,7 +54,7 @@
 void TrialInliner::replaceICStub(const ICEntry& entry, CacheIRWriter& writer,
                                  CacheKind kind) {
   ICFallbackStub* fallbackStub = entry.fallbackStub();
-  fallbackStub->discardStubs(cx(), script_);
+  fallbackStub->discardStubs(cx(), root_->owningScript());
 
   // Note: AttachBaselineCacheIRStub never throws an exception.
   bool attached = false;
@@ -298,6 +302,7 @@
 }
 
 void InliningRoot::trace(JSTracer* trc) {
+  TraceEdge(trc, &owningScript_, "inlining-root-owning-script");
   for (auto& inlinedScript : inlinedScripts_) {
     inlinedScript->trace(trc);
   }